In [2]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import roc_auc_score,accuracy_score
from sklearn.ensemble import BaggingClassifier

In [3]:
df=pd.read_csv("Sonar.csv")
df.head()

V1      V2      V3      V4      V5      V6      V7      V8      V9  \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

      V10  ...     V52     V53     V54     V55     V56     V57     V58  \
0  0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1  0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2  0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
3  0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
4  0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   

      V59     V60  Class  
0  0.0090  0.0032      R  
1  0.0052  0.0044      R  
2  0.0095  0.0078      R  
3  0.0040  0.0117      R  
4  0.0107  0.0094      R  

[5 rows x 61 columns]

In [4]:
X = df.iloc[:,:-1]
y = df.iloc[:,-1]

In [5]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.3,random_state=2022,stratify=y)

In [6]:
log_reg = LogisticRegression()
dtc =  DecisionTreeClassifier(random_state=2022)

In [9]:
model_bg = BaggingClassifier(base_estimator=log_reg,n_estimators=15,max_samples=X_train.shape[0],max_features=X_train.shape[1],random_state=2022,oob_score=True)
model_bg.fit(X_train,y_train)
y_pred_prob = model_bg.predict_proba(X_test)[:,1]
print("roc_auc_score with log_reg  = ",roc_auc_score(y_test, y_pred_prob))

roc_auc_score with log_reg  =  0.7241379310344828


In [11]:
model_bg = BaggingClassifier(base_estimator=dtc,n_estimators=15,max_samples=X_train.shape[0],max_features=X_train.shape[1],random_state=2022,oob_score=True)
model_bg.fit(X_train,y_train)
y_pred_prob = model_bg.predict_proba(X_test)[:,1]
print("roc_auc_score with dtc  = ",roc_auc_score(y_test, y_pred_prob))

roc_auc_score with dtc  =  0.7408722109533469


From Above we can conclude that DecisionTree Classifier with Bagging is the Best Model for the given Data , Having roc_auc_score of  0.7408722109533469